### ISSUE TITLE
GlobalModelRepository glob pattern not finding the files

### ISSUE THREAD

**Issue by danixeee** [textX/textX#239]
I am trying to figure out why validation is not working within _textX-LS_ VS Code extension.

I have noticed that [glob.glob(filename_pattern, **glob_args)](https://github.com/textX/textX/blob/master/textx/scoping/__init__.py#L123) is not finding model files and the reason is that it is looking in a wrong directory. Argument `filename_pattern` is `**/*.config` which is fine if the process is ran from the same directory, but in my case `os.getcwd()` is completely different directory.

Is there a way to specify the project's root in this case, for e.g. "/Users/me/project/**/*.config"? Where we should try to find files by this pattern? What if model files are on different disk partitions? Also, does it make sense to include `os.getcwd()` path in [this error](https://github.com/textX/textX/blob/master/textx/scoping/__init__.py#L125), so we can see at which location it is trying to find files?

Not sure if it's related to #66.

**Comment by goto40** [comment]
Can you try if this helps:
```
global_repo_provider = scoping_providers.FQNGlobalRepo(
   glob_args={"recursive": True})
```
If this helps, we should (I can) make it more explicit on http://textx.github.io/textX/stable/scoping/ ("Arguments can be passed to the glob.glob function (glob_args), e.g., to enable recursive globbing.")

The reason not to include this setting by default is, that some python versions do not provide this feature...

**Comment by danixeee** [comment]
I just debug the [recipec](https://github.com/textX/recipec) example and `{"recursive": True}` is passed to _glob_ function. I think that's fine, but the root directory where _glob_ is trying to find files is location where my server is, instead of my project's location.

Location of server:    `/...../textX/textX-LS/textX-LS/server/`
Location of project:   `/...../textX/recipes/`
_os.getcwd()_:           `/...../textX/textX-LS/textX-LS/server/`
_filename_pattern_:   `**/*.config`
_glob_args_: `{"recursive": True}`

Does it make sense to pass _project_root_ parameter when calling `metamodel.model_from_str(model, file_name=file_path)`? Or maybe multiple project roots where `GlobalModelRepository` will look for files?

**Comment by goto40** [comment]
You are absolutely right, @danixeee.

I understand the problem.
 * I think it may be problematic to use all roots of all model files.
 * Do you call `model_from_str` for many model files during editing? I assume yes.

Maybe passing `project_root` as you suggested is a very good solution (because it is a clear solution)! The bad thing about it is, that we have to pass the project root somehow to the scope provider (e.g. through `model_from_xxx` as you suggested)

Other solutions (more automatic solutions) suffer from unclarity (to me) about where to take the root dir from!

@igordejanovic, if you also have an opinion on this topic, your comments are welcome!

**Comment by goto40** [comment]
For our use case:  is it that we have one python interpreter per editor-project? Or one metamodel object per editor-project? This would allow to use global configuration for the search/project path.

**Comment by danixeee** [comment]
> I think it may be problematic to use all roots of all model files.

I agree, I am not sure how we would automatically determinate a root based on model files, because those could be arbitrary nested.

> Do you call model_from_str for many model files during editing? I assume yes.

Yes.

I didn't take a look at scope provider implementation, so I am not sure what would be the best solution in this case.

@goto40 @igordejanovic What do you think about having `settings.py` module for some global configurations, similar as Django's?

> For our use case: is it that we have one python interpreter per editor-project? Or one metamodel object per editor-project? This would allow to use global configuration for the search/project path.

Only one language server is started which can validate multiple (in)dependent models/projects.

**Comment by danixeee** [comment]
@goto40 I was wrong, global configuration might not be a good solution because we will use one interpreter for multiple projects (at least in textX-LS case for now).

It looks like that **before** searching for models that conforms to a _filename_pattern_ **with wildcards**, we have _at least one_ registered model in our repository. Please take a look at my solution and let me know if that could work and should I create a PR. Note that I haven't dived deeply into the codebase, but `recipec` example is working fine using the _textX-LS_ now. :)

Solution: https://github.com/textX/textX/tree/analysis/issue239

**Comment by goto40** [comment]
I had a look.

I have concerns regarding the determinism: if I open a file a included by another file b in my project, I get different results if file b is already loaded or not.

**main problem**: The problem with the global scope provider is that we need a project root. We do not have an abstraction for that. 

At the moment the current work dir is used for command Line apps. For an editor project we need something more sophisticated...

We need a project or workspace abstraction (like in eclipse)... Then we could open a file in that workspace.

How about adding such a thing (passed as optional argument to model_from_xx). The workspace could just represent the root directory and all opened files (in the first implementation). Different implementations could be provided. 

What do you think?

**Comment by igordejanovic** [comment]
Can we do these and if not why?:
- Have a stateless meta-model, e.g. after meta-model is instantiated it can be used by different projects in the same interpreter (we do keep some global model repo on the meta-model level, do we?).
- Have a single `root` model for a collection of models. Usually this is the first model that is loaded and the rest could be imported by it or loaded afterwards. This `root` model path can then be used as the relative position to search for and load all other model (i.e. the project root folder) and the home of global model repo for the project.

@danixeee I agree to add `os.getcwd()` information to the exception above as it can be helpful in debugging.

**Comment by igordejanovic** [comment]
I'm a bit reluctant to introducing the concept of workspace to the textX directly. Maybe I'm wrong but I would prefer to keep it simple and IDE agnostic.

**Comment by danixeee** [comment]
> I have concerns regarding the determinism: if I open a file a included by another file b in my project, I get different results if file b is already loaded or not.

Could we extend a logic to check if a model is already loaded?

Using the _recipec_ example, if I change `bananentag.plan`, _glob_ function is called with following`filename_pattern` arguments (in exact order):

NOTE: "/.../.../" is absolute path.

- `/.../.../schinken-bananen.recipe`
- `**/*.config`
- `**/*.ingredient`
- `/.../.../bananen-schoko-muffins.recipe`
- `**/*.config`
- `**/*.ingredient`
- `/.../.../teheraner-reis-schuessel.recipe`
- `**/*.config`
-  `**/*.ingredient`

Only the first model is loaded from it's absolute path location and second one fails _with current implementation from the master branch_.

> Have a single root model for a collection of models. Usually this is the first model that is loaded and the rest could be imported by it or loaded afterwards. This root model path can then be used as the relative position to search for and load all other model (i.e. the project root folder) and the home of global model repo for the project.

Could we turn `ModelRepository.filename_to_model` to _OrderedDict_ and be sure that first inserted model is a `root` model? Will directory of that model be a root directory for all other models?

> I'm a bit reluctant to introducing the concept of workspace to the textX directly. Maybe I'm wrong but I would prefer to keep it simple and IDE agnostic.

I agree with this one, actually our language server keeps that information.

**Comment by goto40** [comment]
> I'm a bit reluctant to introducing the concept of workspace to the textX directly. Maybe I'm wrong but I would prefer to keep it simple and IDE agnostic.

I understand. However, loading a subcomponent file (included by a main model file) cannot be loaded in the same way (1) on its own and (2) in combination with the main model file.

Of course this could be handled outside textX.  Outside textX, however, makes it difficult to integrate in an ide.

Note: workspace - in the sense I mean - just represents a base dir (which is cwd in the default use case).

**Comment by goto40** [comment]
@danixeee : I think your problems originate from the lack of a defined project directory.

It is unclear to me, in what extend we should integrate this... I still see a 'workspace' object as an elegant solution (in the sense mentioned above, without too much complexity, just a base directory).

I think we need some time to sleep on it.

**Comment by goto40** [comment]
If we define the first model as the root model, as you suggested @danixeee and @igordejanovic , then we have a similar solution.

However, this would make the idea of a workspace/base directory less explicit. And it does not allow to edit a subcomponent model file without opening the master.

This is OK. As I said we should sleep on it...

**Comment by goto40** [comment]
>... (we do keep some global model repo on the meta-model level, do we?).

Indeed this is possible. By default, we do not do this. In that (default) case, the meta-model is stateless, as far as I see (tbc).

**Comment by goto40** [comment]
If the workspace is too heavy, we could also introduce a `_tx_root_dir` for a model? However, a `_tx_workspace` could in part take over some responsibilities from `_tx_model_repository`.

By default, `_tx_root_dir` could be set to cwd. But it could also be specified via `model_from_xx`.

**Comment by igordejanovic** [comment]
This is solved by meta-model parameters (see #240).
Closed via #243
